File: solutions/arrays-hashing/242-valid-anagram.py
"""
242. Valid Anagram
Easy

Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different
word or phrase, typically using all the original letters exactly once.

Example:
Input: s = "anagram", t = "nagaram"
Output: true
"""

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
Two strings are anagrams if they contain the exact same characters with the same frequencies. We can verify this by counting character frequencies in both strings.

### APPROACH:
1. **Length check**: If strings have different lengths, they can't be anagrams
2. **Count characters**: Use a hash map or array to count frequency of each character
3. **Compare counts**: Both strings should have identical character frequency distributions

### WHY THIS WORKS:
- Anagrams are rearrangements of the same letters
- Character frequency is invariant under rearrangement
- If two strings have the same character frequencies, they must be anagrams

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1) - at most 26 lowercase letters

### MULTIPLE APPROACHES:

#### Approach 1: Counter (Most Concise)
```python
return Counter(s) == Counter(t)
```

#### Approach 2: Manual Counting (More Control)
- Count characters in first string
- Decrement counts for second string
- All counts should be zero

#### Approach 3: Sorting (Alternative)
```python
return sorted(s) == sorted(t)
```
Time: O(n log n), but very simple

### EXAMPLE WALKTHROUGH:
```
s = "anagram", t = "nagaram"

Character counts for s:
a: 3, n: 1, g: 1, r: 1, m: 1

Character counts for t:
n: 1, a: 3, g: 1, r: 1, m: 1

Both have identical counts → True
```

### EDGE CASES:
- Empty strings → True (both empty)
- Different lengths → False immediately
- Single character → direct comparison

</details>

from collections import Counter


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        """
        Approach: Character frequency counter
        Time Complexity: O(n)
        Space Complexity: O(1) - at most 26 characters
        """
        return Counter(s) == Counter(t)

    def isAnagramManual(self, s: str, t: str) -> bool:
        """
        Manual counting approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        if len(s) != len(t):
            return False

        count = {}

        # Count characters in s
        for char in s:
            count[char] = count.get(char, 0) + 1

        # Subtract counts for t
        for char in t:
            if char not in count:
                return False
            count[char] -= 1
            if count[char] < 0:
                return False

        return all(v == 0 for v in count.values())

    def isAnagramSort(self, s: str, t: str) -> bool:
        """
        Sorting approach
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        """
        return sorted(s) == sorted(t)


# Test cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1
    s1, t1 = "anagram", "nagaram"
    print(f"s: '{s1}', t: '{t1}'")
    print(f"Output: {solution.isAnagram(s1, t1)}")  # True

    # Test case 2
    s2, t2 = "rat", "car"
    print(f"s: '{s2}', t: '{t2}'")
    print(f"Output: {solution.isAnagram(s2, t2)}")  # False

File: solutions/arrays-hashing/alternatives/242-valid-anagram.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 * **Step 1:** [description]
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Check if two strings are valid anagrams
 *
 * @param {string} s - First string
 * @param {string} t - Second string
 * @return {boolean} - True if strings are anagrams, false otherwise
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1) - at most 26 lowercase letters
 */
function isAnagram(s, t) {
    // Quick length check
    if (s.length !== t.length) {
        return false;
    }

    // Count character frequencies
    const charCount = {};

    // Count characters in first string
    for (let char of s) {
        charCount[char] = (charCount[char] || 0) + 1;
    }

    // Subtract character counts using second string
    for (let char of t) {
        if (!charCount[char]) {
            return false; // Character not in first string or count exhausted
        }
        charCount[char]--;
    }

    return true;
}

/**
 * Alternative solution using sorting
 *
 * @param {string} s - First string
 * @param {string} t - Second string
 * @return {boolean} - True if strings are anagrams, false otherwise
 *
 * Time Complexity: O(n log n) - due to sorting
 * Space Complexity: O(n) - for sorted strings
 */
function isAnagramSort(s, t) {
    if (s.length !== t.length) {
        return false;
    }

    // Sort both strings and compare
    return s.split('').sort().join('') === t.split('').sort().join('');
}

/**
 * Alternative solution using character frequency array (for lowercase letters only)
 *
 * @param {string} s - First string
 * @param {string} t - Second string
 * @return {boolean} - True if strings are anagrams, false otherwise
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1) - fixed size array of 26
 */
function isAnagramArray(s, t) {
    if (s.length !== t.length) {
        return false;
    }

    // Array to count frequency of each letter (a-z)
    const count = new Array(26).fill(0);

    for (let i = 0; i < s.length; i++) {
        count[s.charCodeAt(i) - 'a'.charCodeAt(0)]++; // Increment for s
        count[t.charCodeAt(i) - 'a'.charCodeAt(0)]--; // Decrement for t
    }

    // All counts should be zero for anagrams
    return count.every(c => c === 0);
}

function solve(s, t) {
    return isAnagram(s, t);
}

/**
 * Test cases for Problem 242: Valid Anagram
 */
function testSolution() {
    console.log('Testing 242. Valid Anagram');

    // Test case 1: Basic anagram - "anagram" and "nagaram"
    const result1 = isAnagram("anagram", "nagaram");
    const expected1 = true;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Not anagram - "rat" and "car"
    const result2 = isAnagram("rat", "car");
    const expected2 = false;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Empty strings
    const result3 = isAnagram("", "");
    const expected3 = true;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Different lengths
    const result4 = isAnagram("a", "ab");
    const expected4 = false;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    // Test case 5: Single character
    const result5 = isAnagram("a", "a");
    const expected5 = true;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);

    // Test case 6: Test sorting approach
    const result6 = isAnagramSort("listen", "silent");
    const expected6 = true;
    console.assert(result6 === expected6, `Test 6 failed: expected ${expected6}, got ${result6}`);

    // Test case 7: Test array approach (lowercase only)
    const result7 = isAnagramArray("evil", "vile");
    const expected7 = true;
    console.assert(result7 === expected7, `Test 7 failed: expected ${expected7}, got ${result7}`);

    console.log('All test cases passed for 242. Valid Anagram!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 242. Valid Anagram ===');
    console.log('Category: Arrays Hashing');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    isAnagram,
    isAnagramSort,
    isAnagramArray,
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on arrays hashing concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
